cmd/agent-smoke/internal/tasks.go

```diff
@@ -0,0 +1,71 @@
+package smoke
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// The Task interface represents units of work performed by some function on
+// some frequency interval by the smoke framework.
+type Task interface {
+	Task() (func(context.Context, *Smoke) error, time.Duration)
+}
```

Comment:
The nested functions and return arguments read a little awkwardly here. WDYT about:

```go
type Task interface {
  Frequency() time.Duration 
  Run(context.Context) error 
}

type deletePodTask struct {
  clientset *kubernetes.Clientset

  namespace string 
  pod string 
  duration time.Duration 
}

func (t *deletePodTask) Frequency() time.Duration { return t.duration } 
func (t *deletePodTask) Run(ctx context.Context) error {
  var zero int64 
  err := t.clientset.CoreV1().Pods(t.namespace).Delete(ctx, t.pod, metav1.DeleteOptions{
    GracePeriod: &zero, 
  })
  if err != nil {
    // caller can log the error and decide whether to abort the task 
    // or continue scheduling it  
    return fmt.Errorf("failed to delete pod %q: %w", t.pod, err) 
  }
  return nil
}
```

Reply:
Alternatively, it might not make sense for the Task to store its frequency at all if it's only purpose is to be returned back to the parent which used to know the value in the first place. 

You could potentially track tasks with their frequency instead:

```go
type Smoke struct {
  tasks []scheduledTask 
}

type scheduledTask struct {
  Task 
  Frequency time.Duration 
}

func NewSmokeTest(/* args */) (*Smoke, error) {
  var s Smoke 
  // init, whatever
 
  s.tasks = append(
     s.tasks, 
     repeatingTask{
        Frequency: /* something */, 
        Task: deletePodTask{namespace: namespace, pod: pod},
     },
  )

  return &s, nil 
}
```

Reply:
I think the repeatingTask struct makes the most sense here, made this update.